applylib/applyset/applyset.go

```diff
@@ -153,11 +242,151 @@ func (a *ApplySet) ApplyOnce(ctx context.Context) (*ApplyResults, error) {
 			results.applyError(gvk, nn, fmt.Errorf("error from apply: %w", err))
 			continue
 		}
-
+		visitedUids.Insert(lastApplied.GetUID())
 		tracker.lastApplied = lastApplied
 		results.applySuccess(gvk, nn)
 		tracker.isHealthy = isHealthy(lastApplied)
 		results.reportHealth(gvk, nn, tracker.isHealthy)
 	}
+
+	// We want to be more cautions on pruning and only do it if all manifests are applied.
+	if a.prune && results.applyFailCount == 0 {
+		klog.V(4).Infof("Prune is enabled")
+		err := func() error {
+			pruneObjects, err := kapplyset.FindAllObjectsToPrune(ctx, a.client, visitedUids)
+			if err != nil {
+				return err
+			}
+			if err = a.deleteObjects(ctx, pruneObjects, results); err != nil {
+				return err
+			}
+			return nil
+		}()
+		if err != nil {
+			klog.Errorf("encounter error on pruning %v", err)
+			if e := a.updateParentLabelsAndAnnotations(ctx, kapplyset, "superset"); e != nil {
```

Comment:
I don't think we need to update the parent labels here?

Reply:
This should be moved to the `BeforeApply`.